### PR TITLE
Functions for merging meshes, minor updates to triangulation

### PR DESCRIPTION
**Pull request by saubhagya-gatech**
- Combines multiple 2D meshes into a single mesh. It is assumed that the meshes to be combined have common vertices on the shared edge (no steiner points). Ids of the entities in the labeled set for each mesh to be merged are also updated and assigned to the combined mesh
- minor updates to triangulation, for example, allowing users to add additional hole points. 
- updated function defining region based on stream order in the quad mesh
- storing starting gid for each quad mesh in the root node properties.

### CODE REVIEW

**Comment by saubhagya-gatech** [comment]
One failing test is related to downloading the NHD dataset. All tests pass on the local machine.

File: watershed_workflow/mesh.py [hunk 1]
```diff
@@ -1565,3 +1565,85 @@ def create_submesh(m2, shp):
...
+
+    # merge conn lists
+    combined_conn = mesh1.conn + adjusted_conn2 
+
+    # merged mesh creation
+    m2_combined = watershed_workflow.mesh.Mesh2D(coords=np.array(combined_coords), conn=combined_conn)
+
+    # trasferring labeled sets
+    if transfer_labeled_sets:
+        m2_combined.labeled_sets = m2_combined.labeled_sets + mesh1.labeled_sets
```
Comment (ecoon) [inline watershed_workflow/mesh.py:1633]: I think this breaks if m1 and m2 both have labeled sets that share the same label.  I'm not sure what to do about this -- probably there needs to be an option.  One could imagine two meshes, both have had land cover labeled sets based on NLCD. -- in that case, the merged mesh with LS = 51 should be the union of the two labeled sets.  Alternatively, one could imagine that the user just used a counter for soil types as the label, and therefore just wants to append the two labeled set lists and make the labels unique.

The key thing is that the merged list of labeled sets must have unique labels.

File: watershed_workflow/regions.py [hunk 2]
```diff
@@ -223,32 +223,25 @@ def add_river_corridor_regions(m2, rivers, labels=None):
     labels: list(str), optional
       List of names, one per river.
     """
-    rivers_mls = [shapely.geometry.MultiLineString(list(river)) for river in rivers]
     if labels is None:
-        labels = []
-        for i, p in enumerate(rivers_mls):
-            label = f'river_corridor {i}'
-            labels.append(label)
+      labels = []
```
Comment (ecoon) [inline watershed_workflow/regions.py:227]: your spacing looks off here?  zapf will fix it.

File: watershed_workflow/regions.py [hunk 3]
```diff
@@ -325,44 +303,42 @@ def add_regions_by_stream_order(m2, river, river_id=0):
...
-def add_region_by_reach_id(m2, rivers, reach_ids=None):
+def add_region_by_reach_id(m2, river, reach_ids=None):
     """Add labeled sets to m2 for reaches of each stream order .
      
     Parameters:
     -----------
     m2: watershed_workflow.mesh.Mesh2D object
       2D mesh elevated on DEMs
-    rivers: list(watershed_workflow.river_tree.RiverTree)
+    river: watershed_workflow.river_tree.RiverTre
```
Comment (ecoon) [inline watershed_workflow/regions.py:313]: missing an e?

**Review by ecoon — CHANGES_REQUESTED** [review]
See notes

File: watershed_workflow/__init__.py [hunk 4]
```diff
@@ -1185,6 +1187,8 @@ def triangulate(hucs,
         List of objects, whose boundary (in the case of
         polygons/waterbodies) or reaches (in the case of River) will
         be present in the edges of the triangulation.
+    hole_points : list(shapely.Point), optional
+        List of points inside the polygons to be left as holes/voids (excluded from mesh)
```
Comment (ecoon) [inline watershed_workflow/__init__.py:1191]: also document refinement_polygons.  I'm not sure what that is?

File: watershed_workflow/__init__.py [hunk 5]
```diff
@@ -1159,6 +1159,8 @@ def triangulate(hucs,
                 rivers=None,
                 river_corrs=None,
                 internal_boundaries=None,
+                hole_points=None,
+                refinement_polygons = None,
```
[the remark this answers is not in the document]
Comment (saubhagya-gatech) [reply watershed_workflow/__init__.py:1163]: Thanks for the zapf recipe

File: watershed_workflow/__init__.py (hunk 4 above)
Comment (saubhagya-gatech) [reply watershed_workflow/__init__.py:1191]: I think that is a depreciated option. I just removed that. One can always define a custom refinement function, right and use polygons or lines in the refinement function as needed?

File: watershed_workflow/mesh.py (hunk 1 above)
Comment (saubhagya-gatech) [reply watershed_workflow/mesh.py:1635]: hmm, it is tricky. We have to internally identify or put a flag if it is NLCD or soil, or broadly speaking, needs to be unionized or extended? Will that go in labeled_set class?

Comment (ecoon) [reply watershed_workflow/mesh.py:1635]: I think the best bet is to compare names first.  If names are the same, then merge them.  Then compare labels.  If the labels are the same and the names are different, then the user probably doesn't care about the label and we can "relabel" one of them.

The reason being that the user will likely also be scripting input spec e.g. soil types, etc, that rely on the name.  And most of our names include some unique identifier (MUKEY, NLCD name, "surface", etc) that really is a classifier and we want them to be merged.

File: watershed_workflow/__init__.py (hunk 4 above)
Comment (ecoon) [reply watershed_workflow/__init__.py:1190]: Well, the case we used to support is the case where we want to have the river be edges in the mesh.  I'm not sure that is really ever helpful though, so maybe I'm ok with losing that option.

File: watershed_workflow/mesh.py (hunk 1 above)
Comment (saubhagya-gatech) [reply watershed_workflow/mesh.py]: Transferring labeledsets option will be added in the future. For now, meshes-to-be-merged should not have labeled-sets. They should be added after merging.

**Review by ecoon — APPROVED** [review]

**Pull request merged**